exist-core/src/main/java/org/exist/xquery/functions/fn/transform/Options.java

```diff
@@ -0,0 +1,662 @@
+import org.w3c.dom.Node;
+
+import javax.annotation.Nullable;
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.Attribute;
+import javax.xml.stream.events.StartElement;
+import javax.xml.stream.events.XMLEvent;
+import javax.xml.transform.Source;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamSource;
+import java.io.Reader;
+import java.io.StringReader;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.*;
+
+import static com.evolvedbinary.j8fu.tuple.Tuple.Tuple;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.exist.Namespaces.XSL_NS;
+import static org.exist.xquery.functions.fn.transform.Options.Option.*;
+import static org.exist.xquery.functions.fn.transform.Transform.SAXON_CONFIGURATION;
+import static org.exist.xquery.functions.fn.transform.Transform.toSaxon;
+
+/**
+ * Read options into class values in a single place.
+ * <p></p>
+ * This is a bit clearer where we need an option several times,
+ * we know we have read it up front.
+ */
+class Options {
+
+    static final javax.xml.namespace.QName QN_XSL_STYLESHEET = new javax.xml.namespace.QName(XSL_NS, "stylesheet");
+    static final javax.xml.namespace.QName QN_VERSION = new javax.xml.namespace.QName("version");
+
+    private static final long XXHASH64_SEED = 0x2245a28e;
+    private static final XXHash64 XX_HASH_64 = XXHashFactory.fastestInstance().hash64();
+
+    private Map<QName, XdmValue> readParamsMap(final Optional<MapType> option, final String name) throws XPathException {
+
+        final Map<net.sf.saxon.s9api.QName, XdmValue> result = new HashMap<>();
+
+        final MapType paramsMap = option.orElse(new MapType(context));
+        for (final IEntry<AtomicValue, Sequence> entry : paramsMap) {
+            final AtomicValue key = entry.key();
+            if (!(key instanceof QNameValue)) {
+                throw new XPathException(fnTransform, ErrorCodes.FOXT0002, "Supplied " + name + " is not a valid xs:qname: " + entry);
+            }
+            if (entry.value() == null) {
+                throw new XPathException(fnTransform, ErrorCodes.FOXT0002, "Supplied " + name + " is not a valid xs:sequence: " + entry);
+            }
+            result.put(Convert.ToSaxon.of((QNameValue) key), toSaxon.of(entry.value()));
+        }
+        return result;
+    }
+
+    final Tuple2<String, Source> xsltSource;
+    final MapType stylesheetParams;
+    final Map<net.sf.saxon.s9api.QName, XdmValue> staticParams;
+    final float xsltVersion;
+    final Optional<AnyURIValue> resolvedStylesheetBaseURI;
+    final Optional<QNameValue> initialFunction;
+    final Optional<ArrayType> functionParams;
+    final Map<net.sf.saxon.s9api.QName, XdmValue> templateParams;
+    final Map<net.sf.saxon.s9api.QName, XdmValue> tunnelParams;
```

Comment:
quality: Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.

Reply:
Re-ordered.